libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+                             AppendStoresFromProxyFunc          proxy_func,
+                             GDestroyNotify                     proxy_return_destroy,
+                             gpointer                           proxy_data)
+{
+  AppendStoresTaskData *data = g_new0 (AppendStoresTaskData, 1);
+
+  data->ka_proxy = g_object_ref (ka_proxy);
+  data->proxy_func = proxy_func;
+  data->proxy_return_destroy = proxy_return_destroy;
+  data->proxy_data = proxy_data;
+
+  return data;
+}
+
+static void
+append_stores_task_data_free (AppendStoresTaskData *data)
+{
+  g_clear_object (&data->ka_proxy);
+
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (AppendStoresTaskData,
+                               append_stores_task_data_free);
+
+static void
+append_stores_task_from_proxy_thread (GTask        *task,
+                                      gpointer      source,
+                                      gpointer      task_data,
+                                      GCancellable *cancellable)
+{
+  AppendStoresTaskData *data = task_data;
+  g_autoptr(GError) local_error = NULL;
+
+  gpointer results = data->proxy_func (data->ka_proxy,
+                                       data->proxy_data,
+                                       cancellable,
+                                       &local_error);
+
+  if (results == NULL)
+    {
+      g_task_return_error (task, g_steal_pointer (&local_error));
+      return;
+    }
+
+  g_task_return_pointer (task, g_steal_pointer (&results), data->proxy_return_destroy);
+}
+
+static void
+append_stores_task_from_proxy (EosDiscoveryFeedKnowledgeAppProxy *ka_proxy,
+                               AppendStoresFromProxyFunc          proxy_func,
+                               GDestroyNotify                     proxy_return_destroy,
+                               gpointer                           proxy_func_data,
+                               GCancellable                      *cancellable,
+                               GAsyncReadyCallback                callback,
+                               gpointer                           user_data)
+{
+  g_autoptr(GTask) task = g_task_new (NULL, cancellable, callback, user_data);
+  g_autoptr(AppendStoresTaskData) data = append_stores_task_data_new (ka_proxy,
+                                                                      proxy_func,
+                                                                      proxy_return_destroy,
+                                                                      proxy_func_data);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_set_task_data (task,
+                        g_steal_pointer (&data),
+                        (GDestroyNotify) append_stores_task_data_free);
+  g_task_run_in_thread (task, append_stores_task_from_proxy_thread);
+}
+
+typedef gpointer (*ElementCopyFunc) (gpointer element);
+
+static GPtrArray *
+flat_map_ptr_arrays (GPtrArray       *ptr_arrays,
```

Comment:
I wonder if we could not base our API around `GList` instead; that would make it much easier/cheaper to concatenate a list of lists.

Reply:
Yeah, I was thinking something along the same lines. GList is O(N) access, but since we access sequentially that should not matter.